src/ucp/stream/stream_recv.c

```diff
@@ -268,38 +272,83 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
                  ucp_ep_h ep, void *buffer, size_t count,
                  ucp_datatype_t datatype, ucp_stream_recv_callback_t cb,
                  size_t *length, unsigned flags)
+{
+    ucp_request_param_t param = {
+        .op_attr_mask   = UCP_OP_ATTR_FIELD_DATATYPE |
+                          UCP_OP_ATTR_FIELD_CALLBACK |
+                          UCP_OP_ATTR_FIELD_FLAGS,
+        .cb.recv_stream = (ucp_stream_recv_nbx_callback_t)cb,
+        .flags          = flags,
+        .datatype       = datatype
+    };
+
+    return ucp_stream_recv_nbx(ep, buffer, count, length, &param);
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nbx,
+                 (ep, buffer, count, length, param),
+                 ucp_ep_h ep, void *buffer, size_t count, size_t *length,
+                 const ucp_request_param_t *param)
 {
     ucs_status_t        status     = UCS_OK;
     ucp_ep_ext_proto_t  *ep_ext    = ucp_ep_ext_proto(ep);
+    uint32_t            flags      = ucp_request_param_flags(param);
+    ucp_datatype_t      datatype;
     size_t              dt_length;
     ucp_request_t       *req;
     ucp_recv_desc_t     *rdesc;
+    uint32_t            attr_mask;
 
     UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_STREAM,
                                     return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
 
-    if (ucs_likely(!UCP_DT_IS_GENERIC(datatype))) {
-        dt_length = ucp_dt_length(datatype, count, buffer, NULL);
-        if (ucs_likely(ucp_stream_recv_nb_is_inplace(ep_ext, dt_length))) {
-            status = ucp_stream_process_rdesc_inplace(ucp_stream_rdesc_get(ep_ext),
-                                                      datatype, buffer, count,
-                                                      dt_length, ep_ext);
-            *length = dt_length;
+    attr_mask = param->op_attr_mask &
+                (UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FLAG_NO_IMM_CMPL);
+    if (ucs_likely(attr_mask == 0)) {
+        datatype = ucp_dt_make_contig(1);
+        if (ucs_likely(ucp_stream_recv_nb_is_inplace(ep_ext, count))) {
+            dt_length = count; /* use dt_lendth to suppress coverity false positive */
+            status    = ucp_stream_process_rdesc_inplace(ucp_stream_rdesc_get(ep_ext),
+                                                         datatype, buffer, count,
+                                                         dt_length, ep_ext);
+            *length = count;
             goto out_status;
         }
+        dt_length = 0;
+    } else if (attr_mask == UCP_OP_ATTR_FIELD_DATATYPE) {
+        datatype  = param->datatype;
+        if (!UCP_DT_IS_GENERIC(datatype)) {
+            dt_length = ucp_dt_length(datatype, count, buffer, NULL);
+            if (ucp_stream_recv_nb_is_inplace(ep_ext, dt_length)) {
```

Comment:
can it be united with default datatype case somehow? Looks very similar

Reply:
yes, looks like, but unification will add branches which is not good